lib/porky_lib/file_service.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require 'aws-sdk-s3'
+require 'singleton'
+
+class PorkyLib::FileService
+  include Singleton
+
+  def read(bucket_name, file_key, options = {})
+    tempfile = Tempfile.new
+    s3.bucket(bucket_name).object(file_key).download_file(tempfile.path, options)
+
+    decrypt_file_contents(tempfile)
+  end
+
+  def write(file, bucket_name, key_id, options = {})
+    return if file.nil? || bucket_name.nil? || key_id.nil?
+
+    data = File.file?(file) ? File.read(file) : file
+    file_key = SecureRandom.uuid
+    tempfile = encrypt_file_contents(data, key_id, file_key)
+
+    perform_upload(bucket_name, file_key, tempfile, options)
+
+    # Remove tempfile from disk
+    tempfile.unlink
+
+    file_key
+  end
+
+  private
+
+  def perform_upload(bucket_name, file_key, tempfile, options)
+    obj = s3.bucket(bucket_name).object(file_key)
+    if options.key?(:metadata)
+      obj.upload_file(tempfile.path, options[:metadata])
+    else
+      obj.upload_file(tempfile.path)
+    end
+  end
+
+  def decrypt_file_contents(tempfile)
+    file_contents = tempfile.read
+
+    # Remove tempfile from disk
+    tempfile.unlink
```

Comment:
What will happen with a massive file in Heroku when we read it into memory?

Is there no option or library report to read S3 files in a stream and then use a stream cipher to enrypt/decrypt to avoid downloading and then loading the whole file into memory?